File: apps/patter-voice-sdk/libraries/typescript/src/tts/cartesia.ts
/** Cartesia TTS for Patter pipeline mode. */
import { CartesiaTTS as _CartesiaTTS } from "../providers/cartesia-tts";

/** Constructor options for the Cartesia `TTS` adapter. */
export interface CartesiaTTSOptions {
  /** API key. Falls back to CARTESIA_API_KEY env var when omitted. */
  apiKey?: string;
  model?: string;
  voice?: string;
  language?: string;
  sampleRate?: number;
  speed?: string | number;
  emotion?: string | string[];
  volume?: number;
  baseUrl?: string;
  apiVersion?: string;
}

/** Options for the carrier-specific factories — same as the constructor minus `sampleRate`. */
export type CartesiaCarrierOptions = Omit<CartesiaTTSOptions, "sampleRate">;

function resolveApiKey(apiKey: string | undefined): string {
  const key = apiKey ?? process.env.CARTESIA_API_KEY;
  if (!key) {
    throw new Error(
      "Cartesia TTS requires an apiKey. Pass { apiKey: '...' } or " +
        "set CARTESIA_API_KEY in the environment.",
    );
  }
  return key;
}

/**
 * Cartesia TTS (sonic-3 GA, ~90 ms TTFB).
 *
 * The default model is `sonic-3` — Cartesia's current GA model. Voice IDs
 * from the previous `sonic-2` family (including the default Katie voice)
 * remain compatible.
 *
 * @example
 * ```ts
 * import * as cartesia from "getpatter/tts/cartesia";
 * const tts = new cartesia.TTS();              // reads CARTESIA_API_KEY
 * const tts = new cartesia.TTS({ apiKey: "..." });
 * ```
 *
 * **Telephony** — use {@link TTS.forTwilio} (PCM @ 16 kHz; the pipeline's
 * carrier-side encoder performs the 16 kHz → 8 kHz + μ-law step itself)
 * or {@link TTS.forTelnyx} (PCM @ 16 kHz, native Telnyx default) on
 * phone calls.
 */
export class TTS extends _CartesiaTTS {
  static readonly providerKey = "cartesia_tts";
  constructor(opts: CartesiaTTSOptions = {}) {
    const key = resolveApiKey(opts.apiKey);
    const { apiKey: _ignored, ...rest } = opts;
    void _ignored;
    super(key, rest);
  }

  /** Pipeline TTS pre-configured for Twilio Media Streams (PCM @ 16 kHz — the pipeline decimates to the 8 kHz wire itself; 8 kHz here was decimated AGAIN → chipmunk audio). */
  static override forTwilio(opts?: CartesiaCarrierOptions): TTS;
  // Parent-compatible overload — accepts the legacy positional form too.
  static override forTwilio(
    apiKey: string,
    options?: Omit<CartesiaTTSOptions, "sampleRate">,
  ): TTS;
  static override forTwilio(
    arg1?: string | CartesiaCarrierOptions,
    arg2?: Omit<CartesiaTTSOptions, "sampleRate">,
  ): TTS {
    const opts: CartesiaCarrierOptions =
      typeof arg1 === "string" ? { apiKey: arg1, ...(arg2 ?? {}) } : (arg1 ?? {});
    return new TTS({ ...opts, sampleRate: 16000 });
  }

  /** Pipeline TTS pre-configured for Telnyx (PCM @ 16 kHz). */
  static override forTelnyx(opts?: CartesiaCarrierOptions): TTS;
  static override forTelnyx(
    apiKey: string,
    options?: Omit<CartesiaTTSOptions, "sampleRate">,
  ): TTS;
  static override forTelnyx(
    arg1?: string | CartesiaCarrierOptions,
    arg2?: Omit<CartesiaTTSOptions, "sampleRate">,
  ): TTS {
    const opts: CartesiaCarrierOptions =
      typeof arg1 === "string" ? { apiKey: arg1, ...(arg2 ?? {}) } : (arg1 ?? {});
    return new TTS({ ...opts, sampleRate: 16000 });
  }
}

File: apps/patter-voice-sdk/libraries/python/getpatter/tts/cartesia.py
"""Cartesia TTS for Patter pipeline mode."""

from __future__ import annotations

import os
from typing import ClassVar, Optional

from getpatter.providers.cartesia_tts import CartesiaTTS as _CartesiaTTS

__all__ = ["TTS"]


def _resolve_api_key(api_key: str | None) -> str:
    key = api_key or os.environ.get("CARTESIA_API_KEY")
    if not key:
        raise ValueError(
            "Cartesia TTS requires an api_key. Pass api_key='...' or "
            "set CARTESIA_API_KEY in the environment."
        )
    return key


class TTS(_CartesiaTTS):
    """Cartesia HTTP TTS (``sonic-3`` GA, ~90 ms TTFB).

    The default model is ``sonic-3`` — Cartesia's current GA model. Voice IDs
    from the previous ``sonic-2`` family (including the default Katie voice)
    remain compatible.

    Example::

        from getpatter.tts import cartesia

        tts = cartesia.TTS()                # reads CARTESIA_API_KEY
        tts = cartesia.TTS(api_key="...", voice="f786b574-...")

    Telephony optimization
    ----------------------
    Use :meth:`for_twilio` (PCM @ 16 kHz (the pipeline decimates to the 8 kHz wire itself), skipping the SDK-side
    16 kHz → 16 kHz (the pipeline decimates to the 8 kHz wire itself) resample before μ-law transcoding) or
    :meth:`for_telnyx` (PCM @ 16 kHz, native Telnyx default) on phone
    calls.
    """

    provider_key: ClassVar[str] = "cartesia_tts"

    def __init__(
        self,
        api_key: str | None = None,
        *,
        model: str = "sonic-3",
        voice: str = "f786b574-daa5-4673-aa0c-cbe3e8534c02",
        language: str = "en",
        sample_rate: int = 16000,
        speed: Optional[str | float] = None,
    ) -> None:
        super().__init__(
            api_key=_resolve_api_key(api_key),
            model=model,
            voice=voice,
            language=language,
            sample_rate=sample_rate,
            speed=speed,
        )

    @classmethod
    def for_twilio(
        cls,
        api_key: str | None = None,
        *,
        model: str = "sonic-3",
        voice: str = "f786b574-daa5-4673-aa0c-cbe3e8534c02",
        language: str = "en",
        speed: Optional[str | float] = None,
    ) -> "TTS":
        """Pipeline TTS pre-configured for Twilio Media Streams (PCM @ 16 kHz (the pipeline decimates to the 8 kHz wire itself)).

        Falls back to ``CARTESIA_API_KEY`` from the env when ``api_key``
        is omitted. See
        :class:`getpatter.providers.cartesia_tts.CartesiaTTS.for_twilio`
        for rationale.
        """
        return cls(
            api_key=_resolve_api_key(api_key),
            model=model,
            voice=voice,
            language=language,
            sample_rate=16000,
            speed=speed,
        )

    @classmethod
    def for_telnyx(
        cls,
        api_key: str | None = None,
        *,
        model: str = "sonic-3",
        voice: str = "f786b574-daa5-4673-aa0c-cbe3e8534c02",
        language: str = "en",
        speed: Optional[str | float] = None,
    ) -> "TTS":
        """Pipeline TTS pre-configured for Telnyx (PCM @ 16 kHz).

        Falls back to ``CARTESIA_API_KEY`` from the env when ``api_key``
        is omitted. See
        :class:`getpatter.providers.cartesia_tts.CartesiaTTS.for_telnyx`
        for the trade-off.
        """
        return cls(
            api_key=_resolve_api_key(api_key),
            model=model,
            voice=voice,
            language=language,
            sample_rate=16000,
            speed=speed,
        )
